src/wallet/wallet.cpp

```diff
@@ -28,12 +28,50 @@
 #include <utilmoneystr.h>
 #include <wallet/fees.h>
 
+#include <algorithm>
 #include <assert.h>
 #include <future>
 
 #include <boost/algorithm/string/replace.hpp>
 
-std::vector<CWalletRef> vpwallets;
+static std::vector<CWallet*> vpwallets;
+
+bool AddWallet(CWallet* wallet)
+{
+    assert(wallet);
+    std::vector<CWallet*>::const_iterator i = std::find(vpwallets.begin(), vpwallets.end(), wallet);
+    if (i != vpwallets.end()) return false;
```

Comment:
μNit: Maybe you could add an
```C++
inline bool HasWallet(CWallet* wallet)
{
    assert(wallet);
    return std::find(vpwallets.begin(), vpwallets.end(), wallet) != vpwallets.end();
}
```
and use that here and below (also removing the assert in both places since it's done inline here).

Reply:
Your suggestion "conflicts" with adding a critical section here. If I do as you say then the lock will be recursive, something we want to avoid AFAICK.

Reply:
That's a pity, but no big deal.